particle-iot/firmware
# Network and cloud diagnostics

### Problem

This PR implements the following diagnostics:

#### Network

- Connection status (ID 8)
- Last connection error (ID 9)
- Number of unanticipated disconnects (ID 12)
- Number of connection attempts (ID 27)
- Disconnection reason (ID 28)

#### Cloud

- Connection status (ID 10)
- Last connection error (ID 13)
- Number of unanticipated disconnects (ID 14)
- Number of connection attempts (ID 29)
- Disconnection reason (ID 30)

### Steps to Test

The testing can be tricky since it can only be done manually in most of the cases. A typical testing procedure for an enum diagnostic data source would be to recreate a number of situations specific to the diagnostic and ensure that all enum values defined by the spec get reported correctly depending on a situation. For the error code diagnostics it's probably easier to fake result codes returned by WLAN/Cellular HAL and the comms library directly in the code.

I'm attaching a simple application that dumps the diagnostic data every 5 seconds to start with.

### Example App

```cpp
#include "application.h"

SYSTEM_THREAD(ENABLED)

namespace {

const Serial1LogHandler logHandler(115200, LOG_LEVEL_WARN, {
    { "app", LOG_LEVEL_ALL }
});

unsigned logMillis = 0;
unsigned logCount = 0;

bool logDiagnosticData(void* appender, const uint8_t* data, size_t size) {
    Log.write(LOG_LEVEL_INFO, (const char*)data, size);
    return true;
}

} // namespace

void setup() {
    logMillis = millis();
}

void loop() {
    if (millis() - logMillis >= 5000) {
        ++logCount;
        Log.printf(LOG_LEVEL_INFO, "%u: ", logCount);
        system_format_diag_data(nullptr, 0, 0, logDiagnosticData, nullptr, nullptr);
        Log.print(LOG_LEVEL_INFO, "\r\n");
        logMillis = millis();
    }
}
```

### References

- [CH8761]

---

### Completeness
